README.md
<h1 align="center"> CodeMastermindHQ </h1>
<!--
<a href="https://github.com/Ajay-Dhangar/CodeMastermindHQ/actions/workflows/codeql.yml"><img src="https://github.com/Ajay-Dhangar/CodeMastermindHQ/actions/workflows/codeql.yml/badge.svg"  alt="CodeQL" /></a> -->

[![Netlify Status](https://api.netlify.com/api/v1/badges/af9f9088-4e9f-433e-b070-2edd6c7e8218/deploy-status)](https://app.netlify.com/)
<a href="https://twitter.com/AJAYDHA27250016"><img src="https://img.shields.io/twitter/follow/AJAYDHA27250016.svg?style=social"  alt="Twitter Follow" /></a>
<a href="https://discord.gg/BFkPKMsfuZ"><img src="https://img.shields.io/discord/102860784329052160.svg" alt="Discord Chat" /></a>
<a href="https://github.com/Ajay-Dhangar/CodeMastermindHQ/stargazers"><img src="https://img.shields.io/github/stars/Ajay-Dhangar/CodeMastermindHQ" alt="Stars Badge"/></a>
<a href="https://github.com/Ajay-Dhangar/CodeMastermindHQ/network/members"><img src="https://img.shields.io/github/forks/Ajay-Dhangar/CodeMastermindHQ" alt="Forks Badge"/></a>
<a href="https://github.com/Ajay-Dhangar/CodeMastermindHQ/pulls"><img src="https://img.shields.io/github/issues-pr/Ajay-Dhangar/CodeMastermindHQ" alt="Pull Requests Badge"/></a>
<a href="https://github.com/Ajay-Dhangar/CodeMastermindHQ/issues"><img src="https://img.shields.io/github/issues/Ajay-Dhangar/CodeMastermindHQ" alt="Issues Badge"/></a>
<a href="https://github.com/Ajay-Dhangar/CodeMastermindHQ/graphs/contributors"><img alt="GitHub contributors" src="https://img.shields.io/github/contributors/Ajay-Dhangar/CodeMastermindHQ?color=2b9348"></a>

[![Contributor Covenant](https://img.shields.io/badge/Contributor%20Covenant-2.1-4baaaa.svg)](CODE_OF_CONDUCT.md)


## Software Development Life Cycle (SDLC) model of CodeMastermindHQ:

1. Requirements Analysis:
   - Purpose: The purpose of the [CodeMastermindHQ](https://codemastermindhq.netlify.app/) project is to provide a platform for learning and practicing coding skills.
   - Objectives: The primary objectives of the project are to offer [coding challenges](https://codemastermindhq.netlify.app/support/), provide [educational resources](https://codemastermindhq.netlify.app/courses), and [foster a community for programmers](https://codemastermindhq.netlify.app/showcase).
   - Stakeholders: The main stakeholders include developers, coding enthusiasts, educators, and learners.
   - Requirements:
     - Functional Requirements: 
       - User registration and authentication system
       - Coding challenges with varying difficulty levels
       - Educational resources such as tutorials, articles, and videos
       - User profile management
       - Community features like forums or chat
     - Non-functional Requirements:
       - Responsive design for seamless user experience across devices
       - High performance and scalability to handle a large user base
       - Secure data storage and transmission
       - Intuitive and user-friendly interface

2. Design:
   - Architecture: The [CodeMastermindHQ](https://codemastermindhq.netlify.app/) project follows a client-server architecture, with a web-based front-end and a backend API.
   - Components:
     - Front-end: HTML, CSS, JavaScript, and frameworks like React for building the user interface.
     - Back-end: Node.js and Express for creating the API, along with a database management system like MongoDB.
   - User Interface: The user interface will be designed to be intuitive, visually appealing, and responsive to different screen sizes.
   - Design Tools: Tools like Figma or Sketch will be used to create wireframes and mockups for the user interface.
   - I used [Docusaurus 2](https://docusaurus.io/) for [CodeMastermindHQ](https://codemastermindhq.netlify.app/).

3. Implementation/Coding:
   - Development Environment: Developers will use IDEs or text editors ([VS Code](https://code.visualstudio.com/)) of their choice along with version control systems like [Git](https://git-scm.com/), [GitHub](https://github.com/).
   - Coding Conventions: Consistent coding conventions will be followed for code readability and maintainability.
   - Technology Stack:
     - Front-end: HTML, CSS, JavaScript, React, Redux, and any additional libraries or frameworks as needed.
     - Back-end: Node.js, Express, MongoDB, and any necessary middleware or libraries.
   - Code Organization: The code will be organized into modules or components for better maintainability and reusability.

4. Testing:
   - Testing Approach: The project will follow a comprehensive testing strategy that includes [unit testing](https://github.com/CodeMastermindHQ/CodeMastermindHQ/blob/main/src/utils/__tests__/jsUtils.test.ts), [integration testing](https://github.com/CodeMastermindHQ/CodeMastermindHQ/tree/main/src/utils), and [user acceptance testing (UAT)](https://github.com/CodeMastermindHQ/CodeMastermindHQ/tree/main/src/utils).
   - Tools and Frameworks: Testing frameworks like [Jest](https://github.com/CodeMastermindHQ/CodeMastermindHQ/tree/main/src/utils) or Mocha will be used for unit and integration testing. UAT will involve real users testing the application.
   - Test Cases: Test cases will be created to cover different scenarios, including positive and negative test cases for each feature.

5. Deployment:
   - Hosting Platform: The project will be hosted on a platform [Netlify](https://www.netlify.com/), which provides easy deployment and scalability.
   - Deployment Process:
     - Set up the hosting environment on [Netlify](https://www.netlify.com/).
     - Configure any necessary environment variables.
     - Automate the build and deployment process using [CI/CD tools](https://github.com/CodeMastermindHQ/CodeMastermindHQ/pull/22#issuecomment-1620943708).
     - Monitor the deployment for any errors or issues.

6. Maintenance and Support:
   - Bug Tracking: Use an issue tracking system like [GitHub Issues](https://github.com/CodeMastermindHQ/CodeMastermindHQ/issues/new) to track and address reported bugs or feature requests.
   - Updates and Enhancements: Regularly update dependencies, frameworks, and libraries to ensure security and compatibility. Continuously enhance the site based on user feedback.
   - Support Channels: Provide a support email or a contact form for users to reach out with any issues or inquiries.
   - Monitoring and Maintenance: Monitor site performance, uptime, and user feedback. Perform regular backups and security updates to ensure the stability and reliability of the site.


CHANGELOG.md
## Changelog

### Version 5.0.1 (2022-04-12)

Create React App 5.0.1 is a maintenance release that improves compatibility with React 18. It also updates templates to use `createRoot` and relaxes the check for older versions of Create React App.

#### Migrating from 5.0.0 to 5.0.1

To migrate from version 5.0.0 to 5.0.1, run the following command inside your project folder:

```
npm install --save --save-exact react-scripts@5.0.1
```

or

```
yarn add --exact react-scripts@5.0.1
```

#### 🐛Bug Fixes

- Fixed an issue with webpack noise being printed only if there was an error or warning.
- Added a warning when not using the latest version of Create React App but without exiting.
- Ensured posix compliant joins for URLs in middleware.

#### 💅Enhancements

- Updated templates to use React 18's `createRoot` method.
- Upgraded the version of RTL (React Testing Library) to support React 18.
- Updated deprecated rules in ESLint configuration.

#### 📝Documentation

- Fixed a typo in the deployment.md file.
- Added the Changelog for version 5.0.0.
- Prevented both npm and yarn commands from being copied in the documentation.

#### 🏠Internal

- Ignored the docs folder when publishing.

### Version 5.0.0 (2021-12-14)

Create React App 5.0.0 is a major release that introduces several new features and upgrades all major dependencies. Here are the highlights:

- Webpack 5
- Jest 27
- ESLint 8
- PostCSS 8
- Fast Refresh improvements and bug fixes
- Support for Tailwind CSS
- Improved package manager detection
- Unpinned all dependencies for better compatibility with other tools
- Dropped support for Node 10 and 12

#### Migrating from 4.0.x to 5.0.0

To migrate from version 4.0.x to 5.0.0, run the following command inside your project folder:

```
npm install --save --save-exact react-scripts@5.0.0
```

or

```
yarn add --exact react-scripts@5.0.0
```

**Note: If you encounter errors after upgrading, you may need to delete your `node_modules` folder and reinstall your dependencies by running `npm install` or `yarn`.**

#### 💥Breaking Changes

- Dropped support for Node 10 and 12.

#### Full Changelog

For a complete list of changes, bug fixes, enhancements, and more, please refer to the full [changelog](https://github.com/CodeMastermindHQ/CodeMastermindHQ/blob/main/CHANGELOG.md) on our GitHub repository.


docusaurus.config.js
// @ts-check
// Note: type annotations allow type checking and IDEs autocompletion

const lightCodeTheme = require('prism-react-renderer/themes/github');
const darkCodeTheme = require('prism-react-renderer/themes/dracula');

const path = require('path');

const npm2yarn = require('@docusaurus/remark-plugin-npm2yarn');
const webpackProxy = require('./src/plugins/webpackProxy/index');
const baseUrl = process.env.BASE_URL || '/';
const isDeployPreview = !!process.env.GITHUB_ACTIONS;

/** @type {import('@docusaurus/types').Config} */
const config = {
  title: 'CodeMastermindHQ',
  tagline: 'Experience the power of real-world coding with CodeMastermindHQ',
  organizationName: 'CodeMastermindHQ', // Usually your GitHub org/user name.
  projectName: 'CodeMastermindHQ', // Usually your repo name.
  baseUrl,
  url: 'https://codemastermindhq.netlify.app/',
  onBrokenLinks: 'throw',
  onBrokenMarkdownLinks: 'warn',
  favicon: 'img/favicon.ico',
  customFields: {
    isDeployPreview,
    description: 'Ajay Dhangar',
  },

  presets: [
    [
      '@docusaurus/preset-classic',
      /** @type {import('@docusaurus/preset-classic').Options} */
      ({
        debug: true,
        docs: {
          path: 'docs',
          //routeBasePath: '/',
          sidebarPath: 'sidebars.js',
          numberPrefixParser: false,
          showLastUpdateAuthor: true,
          showLastUpdateTime: true,
          // Please change this to your repo.
          editUrl: 'https://github.com/CodeMastermindHQ/CodeMastermindHQ/edit/main/',
          remarkPlugins: [npm2yarn, { sync: true }],
        },
        // gtag: !isDeployPreview
        // ? {
        //       trackingID: 'G-M8RTQC8274',
        //       // Optional fields.
        //       anonymizeIP: true, // Should IPs be anonymized?
        //     }
        //   : undefined,
        blog: {
          path: 'blog',
          // Simple use-case: string editUrl
          // editUrl: 'https://github.com/facebook/docusaurus/edit/main/website/',
          // Advanced use-case: functional editUrl
          editUrl: ({ locale, blogDirPath, blogPath, permalink }) =>
            `https://github.com/CodeMastermindHQ/CodeMastermindHQ/edit/main/${blogDirPath}/${blogPath}`,
          editLocalizedFiles: false,
          blogTitle: 'Blog',
          blogDescription: 'Blog',
          blogSidebarCount: 5,
          blogSidebarTitle: 'All our posts',
          routeBasePath: 'blog',
          include: ['**/*.{md,mdx}'],
          exclude: [
            '**/_*.{js,jsx,ts,tsx,md,mdx}',
            '**/_*/**',
            '**/*.test.{js,jsx,ts,tsx}',
            '**/__tests__/**',
          ],
          postsPerPage: 10,
          blogListComponent: '@theme/BlogListPage',
          blogPostComponent: '@theme/BlogPostPage',
          blogTagsListComponent: '@theme/BlogTagsListPage',
          blogTagsPostsComponent: '@theme/BlogTagsPostsPage',
          //remarkPlugins: [require('remark-math')],
          rehypePlugins: [],
          beforeDefaultRemarkPlugins: [],
          beforeDefaultRehypePlugins: [],
          truncateMarker: /<!--\s*(truncate)\s*-->/,
          showReadingTime: true,
          feedOptions: {
            type: 'all',
            title: 'Feed',
            description: 'Feed',
            copyright: '',
            language: `Copyright © ${new Date().getFullYear()} CodeMastermindHQ`,
          },
        },
        pages: {
          remarkPlugins: [npm2yarn],
        },
        theme: {
          customCss: require.resolve('./src/css/custom.css'),
        },
        sitemap: {
          ignorePatterns: ['/tests/**'],
        },
      }),
    ],
  ],

  themes: ['@docusaurus/theme-live-codeblock'],

  themeConfig:
    /** @type {import('@docusaurus/preset-classic').ThemeConfig} */
    ({
      docs: {
        sidebar: {
          hideable: true,
          autoCollapseCategories: true,
        },
      },
      metadata: [
        {
          name: "keywords",
          content:
            "Java, cooking, Express.js, HTML, CSS, MERN stack, MongoDB, Code community, Code mentorship, Code bootcamp, Learning to code, Coding tutorials, Web development tutorials, blog, code, coding, Software development, programming, web development, JavaScript, React,Node.js, Full stack development,DSACodeMastermindHQ, Code Mastermind, GitHub repository,Open-source project, Software development, Programming, CodeMastermindHQ project, Code Mastermind game, Online coding game, Programming puzzle game, JavaScript project, Web development project, GitHub, Vercel, HTML, CSS, JavaScript, Frontend development, Web application, Game development, Logic puzzle, Problem-solving, Coding challenge, Programming practice, Learning to code, Coding education, Web-based game, Browser game",
        },
      ],
      announcementBar: {
        id: 'announcementBar',
        content: `⭐️ If you like CodeMasterMindHQ, give it a star on <a target="_blank" href="https://github.com/CodeMastermindHQ/CodeMastermindHQ">GitHub</a> and join us on <a target="_blank" href="https://www.linkedin.com/groups/14232119/">LinkedIn</a> ⭐️`,

        backgroundColor: '#ff1f21',
        textColor: '#f1f1f1',
        isCloseable: true,
      },
      navbar: {
        hideOnScroll: true,
        title: 'CodeMastermindHQ',
        logo: {
          alt: 'CodeMastermindHQ',
          src: 'img/logo.png',
        },
        items: [
          {
            type: 'doc',
            docId: 'html/basic-level/html-intro',
            label: '📚 Tutorials',
            position: 'left',
          },
          {
            to: '/courses',
            label: '🔍 Courses',
            position: 'left',
            activeBaseRegex: `/courses/`,
          },
          // {
          //   to: '/support',
          //   label: '📓 DSA',
          //   position: 'left',
          //   activeBaseRegex: `/support/`,
          // },

          // {to: 'blog', label: 'Blog', position: 'left'},

          {
            type: 'dropdown',
            label: 'Features',
            position: 'left',
            items: [
              {        
                label: '💻 Live Playground 🚀',         
                to: '/LiveCodeEditor',                
              },
              {

                label: '📓 DSA 🚀',
                to: '/support',
                activeBaseRegex: `/support/`,
              },
              {
                label: '🏷️ Tutorial Tags 🚀',
                to: '/docs/tags/',
                activeBaseRegex: `/docs/tags/`,
              },
              {
                label: '🏷️ Courses Tags 🚀',
                to: '/courses/tags/',
                activeBaseRegex: `/courses/tags/`,
              },             
              {
                html: `<svg class="youtube-icon" xmlns="http://www.w3.org/2000/svg" width="20" height="20" fill="currentColor" viewBox="0 0 16 10">
                    <path d="M8.051 1.999h.089c.822.003 4.987.033 6.11.335a2.01 2.01 0 0 1 1.415 1.42c.101.38.172.883.22 1.402l.01.104.022.26.008.104c.065.914.073 1.77.074 1.957v.075c-.001.194-.01 1.108-.082 2.06l-.008.105-.009.104c-.05.572-.124 1.14-.235 1.558a2.007 2.007 0 0 1-1.415 1.42c-1.16.312-5.569.334-6.18.335h-.142c-.309 0-1.587-.006-2.927-.052l-.17-.006-.087-.004-.171-.007-.171-.007c-1.11-.049-2.167-.128-2.654-.26a2.007 2.007 0 0 1-1.415-1.419c-.111-.417-.185-.986-.235-1.558L.09 9.82l-.008-.104A31.4 31.4 0 0 1 0 7.68v-.123c.002-.215.01-.958.064-1.778l.007-.103.003-.052.008-.104.022-.26.01-.104c.048-.519.119-1.023.22-1.402a2.007 2.007 0 0 1 1.415-1.42c.487-.13 1.544-.21 2.654-.26l.17-.007.172-.006.086-.003.171-.007A99.788 99.788 0 0 1 7.858 2h.193zM6.4 5.209v4.818l4.157-2.408L6.4 5.209z"/>
                  </svg> <label class="youtube-icon-name">Videoes</label>`,
                to: '/YouTubeVideos',
              },
              // {
              //   html: `<svg class="youtube-icon" xmlns="http://www.w3.org/2000/svg" width="20" height="20" fill="currentColor" viewBox="0 0 16 10">
              //       <path d="M8.051 1.999h.089c.822.003 4.987.033 6.11.335a2.01 2.01 0 0 1 1.415 1.42c.101.38.172.883.22 1.402l.01.104.022.26.008.104c.065.914.073 1.77.074 1.957v.075c-.001.194-.01 1.108-.082 2.06l-.008.105-.009.104c-.05.572-.124 1.14-.235 1.558a2.007 2.007 0 0 1-1.415 1.42c-1.16.312-5.569.334-6.18.335h-.142c-.309 0-1.587-.006-2.927-.052l-.17-.006-.087-.004-.171-.007-.171-.007c-1.11-.049-2.167-.128-2.654-.26a2.007 2.007 0 0 1-1.415-1.419c-.111-.417-.185-.986-.235-1.558L.09 9.82l-.008-.104A31.4 31.4 0 0 1 0 7.68v-.123c.002-.215.01-.958.064-1.778l.007-.103.003-.052.008-.104.022-.26.01-.104c.048-.519.119-1.023.22-1.402a2.007 2.007 0 0 1 1.415-1.42c.487-.13 1.544-.21 2.654-.26l.17-.007.172-.006.086-.003.171-.007A99.788 99.788 0 0 1 7.858 2h.193zM6.4 5.209v4.818l4.157-2.408L6.4 5.209z"/>
              //     </svg> <label class="youtube-icon-name">Video PlayLists</label>`,
              //   to: '/docs/tags/',
              //   activeBaseRegex: `/docs/tags/`,
              // }
            ]
          },

          // {
          //   label: '🏷️ Tags',
          //   to: '/courses/tags/',
          //   activeBaseRegex: `/docs/tags/`,
          // },

          { to: 'showcase', label: '🌍 Showcase', position: 'left' },
          {
            href: 'https://github.com/CodeMastermindHQ/CodeMastermindHQ',
            position: 'right',
            className: 'header-github-link',
            'aria-label': 'GitHub repository',
          },
          {
            type: 'dropdown',
            label: '💝 Donation',
            position: 'right',
            items: [
              {
                to: 'https://github.com/sponsors/Ajay-Dhangar',
                label: '💝 Sponsors',
              },
              {
                to: 'https://paytm.business/link/1682044654357/LL_613546699',
                html: '<svg class="paytm-icon" xmlns="http://www.w3.org/2000/svg" viewBox="0 8 25 25" width="85px" height="25px"><path fill="#0d47a1" d="M5.446 18.01H.548c-.277 0-.502.167-.503.502L0 30.519c-.001.3.196.45.465.45.735 0 1.335 0 2.07 0C2.79 30.969 3 30.844 3 30.594 3 29.483 3 28.111 3 27l2.126.009c1.399-.092 2.335-.742 2.725-2.052.117-.393.14-.733.14-1.137l.11-2.862C7.999 18.946 6.949 18.181 5.446 18.01zM4.995 23.465C4.995 23.759 4.754 24 4.461 24H3v-3h1.461c.293 0 .534.24.534.535V23.465zM13.938 18h-3.423c-.26 0-.483.08-.483.351 0 .706 0 1.495 0 2.201C10.06 20.846 10.263 21 10.552 21h2.855c.594 0 .532.972 0 1H11.84C10.101 22 9 23.562 9 25.137c0 .42.005 1.406 0 1.863-.008.651-.014 1.311.112 1.899C9.336 29.939 10.235 31 11.597 31h4.228c.541 0 1.173-.474 1.173-1.101v-8.274C17.026 19.443 15.942 18.117 13.938 18zM14 27.55c0 .248-.202.45-.448.45h-1.105C12.201 28 12 27.798 12 27.55v-2.101C12 25.202 12.201 25 12.447 25h1.105C13.798 25 14 25.202 14 25.449V27.55zM18 18.594v5.608c.124 1.6 1.608 2.798 3.171 2.798h1.414c.597 0 .561.969 0 .969H19.49c-.339 0-.462.177-.462.476v2.152c0 .226.183.396.422.396h2.959c2.416 0 3.592-1.159 3.591-3.757v-8.84c0-.276-.175-.383-.342-.383h-2.302c-.224 0-.355.243-.355.422v5.218c0 .199-.111.316-.29.316H21.41c-.264 0-.409-.143-.409-.396v-5.058C21 18.218 20.88 18 20.552 18c-.778 0-1.442 0-2.22 0C18.067 18 18 18.263 18 18.594L18 18.594z"/><path fill="#00adee" d="M27.038 20.569v-2.138c0-.237.194-.431.43-.431H28c1.368-.285 1.851-.62 2.688-1.522.514-.557.966-.704 1.298-.113L32 18h1.569C33.807 18 34 18.194 34 18.431v2.138C34 20.805 33.806 21 33.569 21H32v9.569C32 30.807 31.806 31 31.57 31h-2.14C29.193 31 29 30.807 29 30.569V21h-1.531C27.234 21 27.038 20.806 27.038 20.569L27.038 20.569zM42.991 30.465c0 .294-.244.535-.539.535h-1.91c-.297 0-.54-.241-.54-.535v-6.623-1.871c0-1.284-2.002-1.284-2.002 0v8.494C38 30.759 37.758 31 37.461 31H35.54C35.243 31 35 30.759 35 30.465V18.537C35 18.241 35.243 18 35.54 18h1.976c.297 0 .539.241.539.537v.292c1.32-1.266 3.302-.973 4.416.228 2.097-2.405 5.69-.262 5.523 2.375 0 2.916-.026 6.093-.026 9.033 0 .294-.244.535-.538.535h-1.891C45.242 31 45 30.759 45 30.465c0-2.786 0-5.701 0-8.44 0-1.307-2-1.37-2 0v8.44H42.991z"/></svg>',
              },
            ]
          },
          {
            type: 'dropdown',
            label: '🌱Community',
            position: 'right',
            items: [
              {        
                // label: 'Discord',         
                href: 'https://discord.com/invite/BFkPKMsfuZ',
                html: `<svg class="discord-icon" stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 640 300" height="20" width="20" xmlns="http://www.w3.org/2000/svg"><path d="M524.531,69.836a1.5,1.5,0,0,0-.764-.7A485.065,485.065,0,0,0,404.081,32.03a1.816,1.816,0,0,0-1.923.91,337.461,337.461,0,0,0-14.9,30.6,447.848,447.848,0,0,0-134.426,0,309.541,309.541,0,0,0-15.135-30.6,1.89,1.89,0,0,0-1.924-.91A483.689,483.689,0,0,0,116.085,69.137a1.712,1.712,0,0,0-.788.676C39.068,183.651,18.186,294.69,28.43,404.354a2.016,2.016,0,0,0,.765,1.375A487.666,487.666,0,0,0,176.02,479.918a1.9,1.9,0,0,0,2.063-.676A348.2,348.2,0,0,0,208.12,430.4a1.86,1.86,0,0,0-1.019-2.588,321.173,321.173,0,0,1-45.868-21.853,1.885,1.885,0,0,1-.185-3.126c3.082-2.309,6.166-4.711,9.109-7.137a1.819,1.819,0,0,1,1.9-.256c96.229,43.917,200.41,43.917,295.5,0a1.812,1.812,0,0,1,1.924.233c2.944,2.426,6.027,4.851,9.132,7.16a1.884,1.884,0,0,1-.162,3.126,301.407,301.407,0,0,1-45.89,21.83,1.875,1.875,0,0,0-1,2.611,391.055,391.055,0,0,0,30.014,48.815,1.864,1.864,0,0,0,2.063.7A486.048,486.048,0,0,0,610.7,405.729a1.882,1.882,0,0,0,.765-1.352C623.729,277.594,590.933,167.465,524.531,69.836ZM222.491,337.58c-28.972,0-52.844-26.587-52.844-59.239S193.056,219.1,222.491,219.1c29.665,0,53.306,26.82,52.843,59.239C275.334,310.993,251.924,337.58,222.491,337.58Zm195.38,0c-28.971,0-52.843-26.587-52.843-59.239S388.437,219.1,417.871,219.1c29.667,0,53.307,26.82,52.844,59.239C470.715,310.993,447.538,337.58,417.871,337.58Z"></path></svg> 
                <label class="discord-icon-name">Discord</label> `,
              },
              {

                // label: 'You Tube',
                href: 'https://www.youtube.com/@Code-Mastermind-HQ',
                html: `<svg class="youtube-icon" xmlns="http://www.w3.org/2000/svg" width="20" height="20" fill="currentColor" viewBox="0 0 16 10">
                    <path d="M8.051 1.999h.089c.822.003 4.987.033 6.11.335a2.01 2.01 0 0 1 1.415 1.42c.101.38.172.883.22 1.402l.01.104.022.26.008.104c.065.914.073 1.77.074 1.957v.075c-.001.194-.01 1.108-.082 2.06l-.008.105-.009.104c-.05.572-.124 1.14-.235 1.558a2.007 2.007 0 0 1-1.415 1.42c-1.16.312-5.569.334-6.18.335h-.142c-.309 0-1.587-.006-2.927-.052l-.17-.006-.087-.004-.171-.007-.171-.007c-1.11-.049-2.167-.128-2.654-.26a2.007 2.007 0 0 1-1.415-1.419c-.111-.417-.185-.986-.235-1.558L.09 9.82l-.008-.104A31.4 31.4 0 0 1 0 7.68v-.123c.002-.215.01-.958.064-1.778l.007-.103.003-.052.008-.104.022-.26.01-.104c.048-.519.119-1.023.22-1.402a2.007 2.007 0 0 1 1.415-1.42c.487-.13 1.544-.21 2.654-.26l.17-.007.172-.006.086-.003.171-.007A99.788 99.788 0 0 1 7.858 2h.193zM6.4 5.209v4.818l4.157-2.408L6.4 5.209z"/>
                  </svg> <label class="youtube-icon-name">You Tube</label>`,
              },
            ]
          },
        ],
      },
      footer: {
        style: 'dark',
        links: [
          {
            title: 'Docs',
            items: [
              {
                label: 'HTML',
                to: '/',
              },
              {
                label: 'JavaScript',
                to: '/',
              },
              {
                label: 'React',
                to: '/',
              },
            ],
          },
          {
            title: 'Courses',
            items: [
              {
                label: 'HTML',
                to: 'courses/html/welcome-html',
              },
              {
                label: 'CSS',
                to: 'courses/',
              },
              {
                label: 'JavaScript',
                to: 'courses/',
              },

              {
                label: 'React',
                to: 'courses/',
              },
            ],
          },
          {
            title: 'Problem Solving',
            items: [
              {
                label: 'Data Structure and Algorithms',
                to: 'support',
              },
              {
                html: `<br>
              <a href="https://app.netlify.com/sites/codemastermindhq/deploys" target="_blank" rel="noreferrer noopener" aria-label="Deploys by Netlify">
                <img src="https://api.netlify.com/api/v1/badges/af9f9088-4e9f-433e-b070-2edd6c7e8218/deploy-status" alt="Deploys by Netlify" width="144" height="35" />
              </a>
            `,
              },
            ],
          },
          {
            title: 'Community',
            items: [
              {
                label: 'Stack Overflow',
                href: 'https://stackoverflow.com/users/18530900/ajay-dhangar',
              },
              {
                label: 'Discord',
                href: 'https://discord.com/invite/BFkPKMsfuZ',
              },
              {
                label: 'Twitter',
                href: 'https://twitter.com/codemastermind2',
              },
              {
                label: 'Discussions',
                href: 'https://github.com/CodeMastermindHQ/CodeMastermindHQ/discussions',
              },
            ],
          },
          {
            title: 'More',
            items: [
              {
                label: 'Blog',
                to: 'blog',
              },
              {
                label: 'GitHub',
                href: 'https://github.com/CodeMastermindHQ/CodeMastermindHQ',
              },
              {
                label: 'LinkedIn Group',
                href: 'https://www.linkedin.com/groups/14232119/',
              },
              {
                label: 'WhatsApp',
                href: 'https://chat.whatsapp.com/HzbYP9F1Sn3ASPW4NCIAkL',
              },

            ],
          },
        ],
        copyright: `Copyright © ${new Date().getFullYear()} CodeMastermindHQ Made with ❤️ by <a href="https://github.com/ajay-dhangar" target="_blank">Ajay-Dhangar</a>`,
      },
      prism: {
        theme: lightCodeTheme,
        darkTheme: darkCodeTheme,
        additionalLanguages: ['powershell'],
        magicComments: [
          {
            className: 'theme-code-block-highlighted-line',
            line: 'highlight-next-line',
            block: { start: 'highlight-start', end: 'highlight-end' },
          },
          {
            className: 'code-block-error-line',
            line: 'This will error',
          },
        ],
      },
    }),
  plugins: [
    './src/plugins/webpackProxy',
    [
      require.resolve('docusaurus-lunr-search'),
      {
        languages: ['en'], // language codes
      },
    ],
    [
      '@docusaurus/plugin-content-docs',
      /** @type {import('@docusaurus/plugin-content-docs').Options} */
      {
        id: 'courses',
        path: 'courses',
        routeBasePath: 'courses',
        sidebarPath: require.resolve('./sidebarsCourses.js'),
        editUrl: 'https://github.com/CodeMastermindHQ/CodeMastermindHQ/edit/documentation/',
        remarkPlugins: [npm2yarn],
        sidebarCollapsible: true,
        numberPrefixParser: false,
        showLastUpdateAuthor: true,
        showLastUpdateTime: true,
      },
    ],
    [
      '@docusaurus/plugin-content-docs',
      /** @type {import('@docusaurus/plugin-content-docs').Options} */
      {
        id: 'support',
        path: './support', // Path to data on filesystem, relative to site dir.
        routeBasePath: 'support', // URL Route.
        include: ['**/*.md'],
        sidebarPath: require.resolve('./sidebarsSupport.js'),
        editUrl: 'https://github.com/CodeMastermindHQ/CodeMastermindHQ/edit/documentation/',
        remarkPlugins: [npm2yarn],
        sidebarCollapsible: true,
        numberPrefixParser: false,
        showLastUpdateAuthor: true,
        showLastUpdateTime: true,
      },
    ],
    [
      '@docusaurus/plugin-ideal-image',
      /** @type {import('@docusaurus/plugin-ideal-image').PluginOptions} */
      {
        quality: 70,
        max: 1030, // max resized image's size.
        min: 640, // min resized image's size. if original is lower, use that size.
        steps: 2, // the max number of images generated between min and max (inclusive)
        disableInDev: false, // By default, the plugin is inactive in development
      },
    ],
    [
      '@docusaurus/plugin-pwa',
      {
        debug: isDeployPreview,
        offlineModeActivationStrategies: [
          'appInstalled',
          'standalone',
          'queryString',
        ],
        // swRegister: false,
        swCustom: path.resolve(__dirname, 'src/sw.js'),
        pwaHead: [
          {
            tagName: 'link',
            rel: 'icon',
            href: 'img/docusaurus.png',
          },
          {
            tagName: 'link',
            rel: 'manifest',
            href: 'manifest.json',
          },
          {
            tagName: 'meta',
            name: 'theme-color',
            content: 'rgb(37, 194, 160)',
          },
          {
            tagName: 'meta',
            name: 'apple-mobile-web-app-capable',
            content: 'yes',
          },
          {
            tagName: 'meta',
            name: 'apple-mobile-web-app-status-bar-style',
            content: '#000',
          },
          {
            tagName: 'link',
            rel: 'apple-touch-icon',
            href: 'img/docusaurus.png',
          },
          {
            tagName: 'link',
            rel: 'mask-icon',
            href: 'img/docusaurus.png',
            color: 'rgb(62, 204, 94)',
          },
          {
            tagName: 'meta',
            name: 'msapplication-TileImage',
            content: 'img/docusaurus.png',
          },
          {
            tagName: 'meta',
            name: 'msapplication-TileColor',
            content: '#000',
          },
        ],
      },
    ],
  ],
};

module.exports = config;
